### PR TITLE
Track Exception-level exceptions in integrations

### LINKED ISSUE #174: ⌨️ What errors do we want to track?
Do we want to track `SyntaxErrors`, `NotImplementedError` en `LoadError`s (etc) in AppSignal? We don't do that now for a lot of implementations, because we only track `StandardError`s, see #168 for fixing mis-usage and my upcoming blog post.

It can however be a good addition to track more errors than those subclassed from `StandardError`. What do you think?

See https://gist.github.com/tombruijn/cad99e541231ba7283f4467dbf68da2b for more details about which error does what + the docs http://ruby-doc.org/core-2.3.1/Exception.html

Edit: As requested:

- https://app.intercom.io/a/apps/yzor8gyw/inbox/conversation/7028024986
- https://app.intercom.io/a/apps/yzor8gyw/respond/inbox/conversation/11111116983
- https://app.intercom.io/a/apps/yzor8gyw/respond/inbox/540709/conversations/11598818180

## TODO

- [x] Review #356 
- [x] Ship an alpha release
- [x] Test the alpha release on staging https://github.com/appsignal/appsignal-server/pull/2543
- [x] Get users to try out the new Exception handling (see Intercom)
- [x] Add automatic intercom message to people upgrading from `< 2.4.0` to `>= 2.4.0` about gem version 2.4 capturing a lot more errors - written, now to enable once blog post is published

### PR DESCRIPTION
**Pull request by tombruijn**
This allows us to track more Exceptions in Ruby without breaking the normal Exception handling in Ruby. The main thing to keep in mind while rescuing Exception is that you re-raise the Exception in case it's an `SystemExit` or `SignalException`, otherwise we might break the normal shutdown behavior of Ruby processes.

Closes #174 

# About this PR

This PR is the public version of this change. I'll create a separate branch and PR for the final change with a collection of neatly rebased commits instead of this PR where I'll just keep adding more commits to.

TLDR: Do not merge this PR, I'll rebase it and create a new one for merging instead.

# DO NOT REBASE/FORCE-PUSH THIS BRANCH!

please 😇  (it's protected on GitHub, but you know, watch it).

### CODE REVIEW

**Comment by tombruijn** [comment]
This PR is continued in #356. I've squashed some commits that are irrelevant to the change and rebased on the `develop` branch. Fixed some conflicts after merging #348.
This PR will be closed. Please check out #356 instead.
🚨  **Warning**: Do not rely on this branch anymore. I'll remove it in the coming days.
